### ISSUE TITLE
Hessian Approximation Test Failed when using Product Manifold with Special Orthogonal Group

### ISSUE THREAD

**Issue by I0x0I** [pymanopt/pymanopt#219]
Hi, I've noticed a problematic hessian when using product manifold with special orthogonal group.

The code I used for testing is enclosed below, my pymanopt version is [2.0.1](https://github.com/pymanopt/pymanopt/releases/tag/2.0.1).

```python
#Adapted From https://github.com/pymanopt/pymanopt/blob/master/tests/manifolds/test_product.py
import autograd.numpy as np

import pymanopt
from pymanopt.manifolds import Euclidean, Grassmann, Product, Sphere, SpecialOrthogonalGroup


m = 100
n = 50

#manifold = Product([Euclidean(m, n), Sphere(n)])
manifold = Product([Euclidean(m, n), SpecialOrthogonalGroup(m,k=n)])

point = manifold.random_point()

@pymanopt.function.autograd(manifold)
def cost(*x):
    return np.sum(
        [np.linalg.norm(a - b) ** 2 for a, b in zip(x, point)]
    )

problem = pymanopt.Problem(manifold, cost)

#Adapted From https://github.com/pymanopt/pymanopt/blob/master/tests/manifolds/_manifold_tests.py
import pymanopt.tools.diagnostics

*_, (slope, *_) = pymanopt.tools.diagnostics.check_directional_derivative(problem)
print("gradient_approximation_test passed? ", 1.95 <= slope <= 2.05)

_, error, _, (slope, *_) = pymanopt.tools.diagnostics.check_directional_derivative(problem, use_quadratic_model=True)
print("hessian_approximation_test passed? ", np.allclose(np.linalg.norm(error), 0) or (2.95 <= slope <= 3.05))
```

**Comment by nkoep** [comment]
Yes, unfortunately the second-order approximation tests are very fickle and highly dependent on the random seed. I tried explicitly fixing the seed for each test, but this still results in inconsistent test behavior across different python version. Unless @NicolasBoumal or @kellertuer have an idea how to make the test more robust, it's probably best to remove it again. One option would be to draw more samples, i.e., run the same test a few times with different realizations and only fail it if the majority of draws fail the slope condition. That feels like a bandaid more than anything else though. It'd be a shame to remove the test altogether because it did reveal one or two bugs in our code in the past.

**Comment by kellertuer** [comment]
I have not yet worked with the second order tests (that is they are also not yet available in Julia); so I can not help much here.

**Comment by NicolasBoumal** [comment]
Hello,
* The math behind the second-order test is described in Section 6.8 of [my book](https://www.nicolasboumal.net/book/): mentioning this here, just in case you want to try to write your own implementation / to compare against the PyManopt implementation.
* The code for the second-order test is implemented the Matlab version as [checkhessian](https://github.com/NicolasBoumal/manopt/blob/master/manopt/tools/checkhessian.m): maybe that's also useful to compare to.
* In the Matlab version, we do not claim success/failure for the test (because that's not easy to do). Rather, we display all relevant diagnostics information (as a plot and also with text output to the console). It is for the user to consider all of that information and decide if the test is successful or not.
* I imagine you checked already, but: did the first-order test pass? Also, there's a pitfall that sometimes causes issues with the rotation factory, namely, the fact that tangent vectors are represented just based on their skew-symmetric part, whereas vectors in the ambient space are represented "as is". Now, I see here you are using automatic differentiation; maybe there is some trouble there. That could be tested on a separate, simple example involving only the rotations factory; then also testing the product manifold tool against AD.
Nicolas

**Comment by nkoep** [comment]
Thanks for the quick feedback, guys! Re your points, @NicolasBoumal:
- Our test is directly based on manopt's `checkhessian` function. AFAICT the only difference to the procedure described in Section 6.8 of your book is that we do not try to find an approximate critical point when the retraction is not 2nd order; we always pick one at random.
- We also include a 1st order test (again based on manopt's implementation). I believe this test is somewhat more stable, but I seem to recall that we also had sporadic failures there when we pick a "bad" point for the Taylor expansion depending on the PRNG state.
- As for the rotation group, the mismatch in representation of tangent vectors was precisely one of the bugs we found after adding the test.

It's a shame, but I believe it's best to remove the test again. In theory we could mark it as an expected failure, but that doesn't exactly feel clean. Thanks for your input :bow:

**Comment by NicolasBoumal** [comment]
Thanks for the follow up!

> we do not try to find an approximate critical point when the retraction is not 2nd order; we always pick one at random.
Same: in Matlab, we pick a point at random, and print a message to tell the user that they may want to find a critical point first (by running any optimization algorithm, a first-order one for example).

These dependencies on random state are strange. It would be useful to know if the problem also occurs with other manifolds, because currently both the rotations factory and the product manifold are suspicious ingredients.

**Comment by I0x0I** [comment]
Hi everyone, thanks for all the help!

I do suspect the Hessian of product manifold and special orthogonal group is problematic.

I have a python and Matlab implementation of my application (R^3xSO3), they have the same starting point and trust regions optimizer parameters. PyManopt and Manopt produce the same f(x) and |grad|, but pyManopt shows a different convergence behavior (many consecutive TR-). They behave similarly when using steepest descent, or solving R^3xS^3.

**Comment by nkoep** [comment]
> Hi everyone, thanks for all the help!
> 
> I do suspect the Hessian of product manifold and special orthogonal group is problematic.
> 
> I have a python and Matlab implementation of my application (R^3xSO3), they have the same starting point and trust regions optimizer parameters. PyManopt and Manopt produce the same f(x) and |grad|, but pyManopt shows a different convergence behavior (many consecutive TR-). They behave similarly when using steepest descent, or solving R^3xS^3.

I doubt there is an issue in the product manifold itself as it merely delegates calls to the product elements. SO(n) is another matter, although I went over the implementation several times already. In any case, I'll merge a PR to remove the approximation tests soon since they cause too many issues at the moment.
